### ISSUE TITLE
Migration from v2.2 to v3 breaks mypy typing of `my_flow.deploy()`

### ISSUE THREAD

**Issue by benjamincerigo** [PrefectHQ/prefect#15275]
### Bug summary

After migrating from v2.2 to v3 there is an error for mypy due to the typing of the return of `my_flow.deploy(...)`.

It would seem that the mypy has changed how it types the return value of `some_flow.deploy(...)` between prefect v2.2 and v3. In v2.2 it is being typed as `Coroutine` and in v3 is being typed as `UUID`. 

Here is the [MR](https://gitlab.com/howtobuildup/phoenix/-/merge_requests/270) for an open-source project that uses prefect that makes the version [update](https://gitlab.com/howtobuildup/phoenix/-/merge_requests/270#note_2095283604).

[Error out CI](https://gitlab.com/howtobuildup/phoenix/-/jobs/7771281402#L281):  "Incompatible return value type (got "list[UUID]", expected "list[Coroutine[Any, Any, Any]]")  [return-value]"

In the code base, we are running the deployment of flows in parallel. This is done by collecting the returns of `some_flow.deploy()` and then using them in `await asyncio.gather(deployment_coroutines)`. [An example of the `some_flow.deploy(...)`](https://gitlab.com/howtobuildup/phoenix/-/blob/phiphi/prefect_v3/python/projects/phiphi/phiphi/health_check_flows.py?ref_type=heads#L125-136) and [the example of `asyncio.gather(deployment_coroutines)`](https://gitlab.com/howtobuildup/phoenix/-/blob/phiphi/prefect_v3/python/projects/phiphi/phiphi/prefect_deployments.py?ref_type=heads#L78-89).

My proposed solution to it is [here](https://gitlab.com/howtobuildup/phoenix/-/merge_requests/270/diffs?commit_id=97620c1bc0fd2d538e7b7953cbb3493f19d6c800). Could you recommend if this is the best option?

I also think that this is a breaking change and should be documented. 

 

### Version info (`prefect version` output)

```Text
Version:             3.0.0
API version:         0.8.4
Python version:      3.11.8
Git commit:          c40d069d
Built:               Tue, Sep 3, 2024 11:13 AM
OS/Arch:             linux/x86_64
Profile:             default
Server type:         cloud
Pydantic version:    2.9.0
```

**Comment by desertaxle** [comment]
Yeah, typing on `@sync_compatible` has always been lacking because its behavior is runtime-dependent. It was updated in https://github.com/PrefectHQ/prefect/pull/14226 to default to sync typing since that was determined to be the more common use case. Challenges like this are a motivating factor for moving towards explicit sync and async interfaces.

I like your idea of adding a note to the docstring for `sync_compatible`. Would you be willing to submit a PR adding that note?

**Comment by benjamincerigo** [comment]
I would vote for explicit sync and async interfaces. In my opinion, transparency around async code makes this easier.

MR is open :)
